File: 1456-maximum-number-of-vowels-in-a-substring-of-given-length/1456-maximum-number-of-vowels-in-a-substring-of-given-length.cpp
class Solution {
public:
    int maxVowels(string s, int k) {
        return myApproach(s,k);
    }

    bool isVowel(char c) {
        static const unordered_set<char> vowels = {'a', 'e', 'i', 'o', 'u'};
        return vowels.find(c) != vowels.end();
    }
    int myApproach(string s, int k) {
        int n = s.length();
        if (n < k) return -1;

        //first pass through window
        int max_vowel = 0;
        for (int i = 0; i < k; i++) {
            if(isVowel(s[i])) max_vowel++;
        }
        int current_vowel = max_vowel;

        //sliding window
        for (int i = k; i < n; i++) {
            if(isVowel(s[i])) current_vowel++;
            if(isVowel(s[i - k])) current_vowel--;
            max_vowel = max(current_vowel, max_vowel);
        }

        return max_vowel;
    }
};

File: 1456-maximum-number-of-vowels-in-a-substring-of-given-length/1456-maximum-number-of-vowels-in-a-substring-of-given-length.py
class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        vowels = {'a', 'e', 'i', 'o', 'u'}  # Set of vowels
        maxVowel = 0
        curVowel = 0
        
        # Initialize the first window
        for i in range(k):
            if s[i] in vowels:
                curVowel += 1
        maxVowel = curVowel
        
        # Slide the window across the string
        for i in range(k, len(s)):
            # Add the new character to the window
            if s[i] in vowels:
                curVowel += 1
            # Remove the character that's sliding out of the window
            if s[i - k] in vowels:
                curVowel -= 1
            
            # Update the maximum vowel count
            maxVowel = max(maxVowel, curVowel)
        
        return maxVowel
            